### PR TITLE
Implement Wrapped gaussian process regression

### PR DESCRIPTION
**Pull request by arthurPignet**
This PR implements Wrapped gaussian process regression on Riemannian Manifolds.

@Tom-Szwagier

Original paper: https://openaccess.thecvf.com/content_cvpr_2018/papers/Mallasto_Wrapped_Gaussian_Process_CVPR_2018_paper.pdf

```bibtex
@INPROCEEDINGS{,
  author={Mallasto, Anton and Feragen, Aasa},
  booktitle={2018 IEEE/CVF Conference on Computer Vision and Pattern Recognition}, 
  title={Wrapped Gaussian Process Regression on Riemannian Manifolds}, 
  year={2018},
  volume={},
  number={},
  pages={5580-5588},
  doi={10.1109/CVPR.2018.00585}}

```

To do:

- [x] Code base
- [x] Tests
- [x] Doc

### CODE REVIEW

**Comment by ninamiolane** [comment]
Excellent! I will not review the code before you ask us to. But it looks good so far!

Btw, this could be included as a notebook in our challenge: https://github.com/geomstats/challenge-iclr-2022 ;)
This would be an excellent submission actually @arthurPignet @Tom-Szwagier 

You can merge the main code in geomstats here, and you could reference the Pull Request in the challenge!

**Comment by arthurPignet** [comment]
> Excellent! I will not review the code before you ask us to. But it looks good so far!
> 

Thank you ! We are planning to have the code ready for review in the next days, so we will probably contact you about it on Monday. 

> Btw, this could be included as a notebook in our challenge: https://github.com/geomstats/challenge-iclr-2022 ;) This would be an excellent submission actually @arthurPignet @Tom-Szwagier
> 
> You can merge the main code in geomstats here, and you could reference the Pull Request in the challenge!

We were not aware of this challenge but we will be glad to contribute! We will start to write such a notebook next week. 
As we are fairly new to this domain, we will probably struggle a bit to find relevant open access real world datasets  to apply our module.

**Comment by ninamiolane** [comment]
Makes sense! Some open-source datasets are available here: 
https://github.com/geomstats/geomstats/blob/master/geomstats/datasets/utils.py
see this notebook for how to load them:
https://github.com/geomstats/geomstats/blob/master/notebooks/01_data_on_manifolds.ipynb

but I understand that they might not be appropriate. Feel free to reach out if you need help to find other datasets!

**Comment by arthurPignet** [comment]
Hello @ninamiolane, I think we are ready for your review! 
Moreover we need your approval in order to start all the workflows. I did not manage to install your pre-commit hook on my laptop, so I am not sure that we will pass all tests directly.

**Comment by arthurPignet** [comment]
We changed the code in order to be compatible with the pytorch and tensorflow backends. I finally black the two files so normally all the workflows will success. I think that we are ready for reviews!

Just a question about black, it keeps ask me to reformat files I didn't modify, is it normal ?

File: geomstats/learning/wrapped_gaussian_process.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
+r"""Wrapped Gaussian Process.
+
+Lead author: .
```
Comment (ninamiolane) [inline geomstats/learning/wrapped_gaussian_process.py:3]: You should put your names here ;)

File: geomstats/learning/wrapped_gaussian_process.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    copy_X_train : bool, default=True
+        If True, a persistent copy of the training data is stored in the
+        object. Otherwise, just a reference to the training data is stored,
+        which might cause predictions to change if the data is modified
+        externally.
+    random_state : int, RandomState instance or None, default=None
+        Determines random number generation used to initialize the centers.
+        Pass an int for reproducible results across multiple function calls.
+
+    [1] Mallasto, A. and Feragen, A. Wrapped gaussian process
```
Comment (ninamiolane) [inline geomstats/learning/wrapped_gaussian_process.py:85]: Could you use this format for the reference?
https://github.com/geomstats/geomstats/blob/60dedbc6bff239016c389b5439aff9dcb87cf6dd/geomstats/geometry/stiefel.py#L209

It allows to show up nicely on the documentation website:
https://geomstats.github.io/geometry.html#module-geomstats.geometry.stiefel

File: geomstats/learning/wrapped_gaussian_process.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            kernel=kernel,
+            alpha=alpha,
+            optimizer=optimizer,
+            n_restarts_optimizer=n_restarts_optimizer,
+            normalize_y=False,
+            copy_X_train=copy_X_train,
+            random_state=random_state,
+        )
+
+        # I know that inheritance seems more appropriate here, but the issue is
```
Comment (ninamiolane) [inline geomstats/learning/wrapped_gaussian_process.py:123]: Sounds good! Could you just remove this comment?

File: geomstats/learning/wrapped_gaussian_process.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+        Returns
+        -------
+        self : object
+            WrappedGaussianProcessRegressor class instance.
+        """
+        if not gs.all(self.space.belongs(y)):
+            raise AttributeError("The target values must belongs to the given space")
+
+        # compute the tangent dataset using the prior
```
Comment (ninamiolane) [inline geomstats/learning/wrapped_gaussian_process.py:168]: Could you move these comments into the description of the docstring of `fit`? You can write something like:

The Wrapped Gaussian process is fit through the following steps:
 1. Compute the tangent dataset using the prior
 2. Fit a Gaussian process regression on the tangent dataset
 3. ...

Putting this in the docstring allows it to show up in the documentation website, and it is still visible from the code.

File: geomstats/learning/wrapped_gaussian_process.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        Returns
+        -------
+        y_samples : ndarray of shape (n_samples_X, n_samples), or \
+            (n_samples_X, n_targets, n_samples)
+            Values of n_samples samples drawn from wrapped Gaussian process and
+            evaluated at query points.
+        """
+        tangent_samples = self._euclidean_gpr.sample_y(X, n_samples, random_state)
+        tangent_samples = gs.cast(tangent_samples, dtype=X.dtype)
+        # flatten the samples
```
Comment (ninamiolane) [inline geomstats/learning/wrapped_gaussian_process.py:269]: Same comments

File: tests/tests_geomstats/test_wrapped_gaussian_process.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        """Test the predict method"""
+        wgpr = WrappedGaussianProcess(
+            self.sphere, metric=self.sphere.metric, prior=self.prior, kernel=self.kernel
+        )
+        wgpr.fit(self.X_sphere, self.y_sphere)
+        y, std = wgpr.predict(self.X_sphere, return_tangent_std=True)
+        self.assertAllClose(std, gs.zeros(*std.shape), atol=1e-4)
+        self.assertAllClose(y, self.y_sphere, atol=1e-4)
+
+    @geomstats.tests.np_and_torch_only
```
Comment (ninamiolane) [inline tests/tests_geomstats/test_wrapped_gaussian_process.py:62]: What is failing with the other backends?

**Review by ninamiolane — COMMENTED** [review]
Fantastic, thank you! Really clean code 🙌 I made some comments, but they are very minor. Congratulations, we are very close to merging this.

File: geomstats/learning/wrapped_gaussian_process.py (hunk 1 above)
Comment (arthurPignet) [reply geomstats/learning/wrapped_gaussian_process.py:3]: Done

File: geomstats/learning/wrapped_gaussian_process.py (hunk 2 above)
Comment (arthurPignet) [reply geomstats/learning/wrapped_gaussian_process.py:85]: Done

File: geomstats/learning/wrapped_gaussian_process.py (hunk 3 above)
Comment (arthurPignet) [reply geomstats/learning/wrapped_gaussian_process.py:123]: done

File: geomstats/learning/wrapped_gaussian_process.py (hunk 5 above)
Comment (arthurPignet) [reply geomstats/learning/wrapped_gaussian_process.py:269]: I added these comments in the docstring

File: tests/tests_geomstats/test_wrapped_gaussian_process.py (hunk 6 above)
Comment (arthurPignet) [reply tests/tests_geomstats/test_wrapped_gaussian_process.py:62]: The samples_y method wasn't working at the time (with a for loop across the samples) with the tf backend. But now that the computation of the exp is vectorized, it is supposed to work. I forgot to remove the decorator, I fix it.

**Comment by arthurPignet** [comment]
> Fantastic, thank you! Really clean code 🙌 I made some comments, but they are very minor. Congratulations, we are very close to merging this.

Thank you for your comments! I took them into account, so the PR should be ready for a second round of review.

**Pull request merged**